Time Selection Feature

ravishanker opened the pull request:
Implemented time selection for blogging reminders

Fixes #15123 

<img width=320 src=https://user-images.githubusercontent.com/990349/128148139-33f88c86-1e27-456c-a5a2-6698b2a47850.png />


To test:

- Go to site settings
- Find **Blogging reminders** setting and click on the item below it
- It should launch select days bottom sheet 
- bottom sheet should now have Notification time item, click on time
- It should show time selection dialog
- Select a time, AM, PM and time should reflect the selection
- De-select all days, time list item should disappear, then select a day, time should appear again

## Regression Notes
1. Potential unintended areas of impact
Reminder scheduling

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing

3. What automated tests I added (or what prevented me from doing so)
Unit testing

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ravishanker committed:
Time Selection Feature

Implemented time selection for blogging reminders

ravishanker committed:
Fix lint errors

ravishanker committed:
Fix failing tests

ravishanker committed:
fix ktlint issues

ravishanker committed:
fix ciktlint issues

ravishanker commented:
Needs an update to FluxC in a separate PR [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2085)

ravishanker committed:
Update BloggingRemindersModelMapper.kt

renanferrari reviewed (commented):
Thanks for the changes! I'm not sure if this is ready for review, but I've taken a quick look and it seems that the time picker is behaving a bit weirdly after I dismiss it by tapping on the outside:


https://user-images.githubusercontent.com/830056/128782688-30170e11-68cd-44dd-a578-87256b1fac40.mp4

Also, I noticed that the time formatting is not following my current locale?

Let me know if this is ready for a proper review, so I can take a deeper look at the code as well.

ravishanker commented:
> Thanks for the changes! I'm not sure if this is ready for review, but I've taken a quick look and it seems that the time picker is behaving a bit weirdly after I dismiss it by tapping on the outside:
> video.7.mp4
> 
> Also, I noticed that the time formatting is not following my current locale?
> 
> Let me know if this is ready for a proper review, so I can take a deeper look at the code as well.

Thanks @renanferrari for a quick look.  Good catch on the issue.  Looks like if the time picker is cancelled, then it is not dismissing the dialog.  

Unfortunately, I couldn't use material MaterialTimePicker even after upgrading material version.  MaterialTimePicker is crashing due to bridge theme in use in the app.

I've now fixed the issue with Cancel of TimePicker.   You can may be review this after  I merge the [FluxC](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2085) PR first, then update this one.  Thanks

ravishanker committed:
Merge branch 'develop' into feature/15123-blogging-remindertime-selection

ravishanker committed:
fix time picker cancellation

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderBottomSheetFragment.kt:
```diff
@@ -65,6 +65,11 @@ class BloggingReminderBottomSheetFragment : BottomSheetDialogFragment() {
                     primaryButton.visibility = View.GONE
                 }
+
+                if (uiState?.timePicker == true) {
+                    val fragment = BloggingReminderTimePicker.newInstance(0, 0)
+                    fragment.show(requireActivity().supportFragmentManager, BloggingReminderTimePicker.TAG)
+                }
```
Instead of using the state to indicate if the dialog should be shown, I believe we could use a separate event, similar to how we do for the blogging reminder bottom sheet itself. You can take a look at the [`BloggingReminderUtils::observeBottomSheet`](https://github.com/wordpress-mobile/WordPress-Android/blob/a2034122225404ea3c567b144549ff1dd888dae6/WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderUtils.kt#L11) method to get an idea of how this would look like.

ravishanker replied:
My first implementation was using a separate event, then changed it to the current one, assuming we would show timepicker in the next screen like iOS and not on top of day selection.  Given it is viewholders and recylerview, it is playing up with the events.  I'll revert to the separate event and see if it works better

ravishanker replied:
Update to implement as per your suggestion.  Thanks

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderTimePicker.kt:
```diff
@@ -0,0 +1,66 @@
+                minute,
+                is24HrFormat)
+
+        timePicker.setCanceledOnTouchOutside(false)
+
+        timePicker.setOnShowListener {
+            timePicker.getButton(BUTTON_NEGATIVE).setOnClickListener {
+                timePicker.dismiss()
+                viewModel.onCancelTimeSelection()
+            }
+        }
```
Unless I'm missing something, I think we could use `setOnDismissListener` here instead, as that would probably cover all the cases we need.

ravishanker replied:
Tried that one and also other listeners,  The OK and Cancel button on the dialog seem to fire the same onTimeset / onTimeChange callbacks.  MaterialTimePicker has all these distinct listeners, unfortunately can't use it.

renanferrari replied:
>The OK and Cancel button on the dialog seem to fire the same onTimeset / onTimeChange callbacks

Hmm 🤔 I actually couldn't reproduce that. I commented out all the lines above and the picker seems to only fire the `OnTimeSetListener` when I click OK. Maybe I'm missing something?

ravishanker replied:
I mean cancel wasn't working properly before I changed to utils implementation.  Cleaned up everything now

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderTimePicker.kt:
```diff
@@ -0,0 +1,66 @@
+        private const val ARG_HOUR = "arg_hour"
+        private const val ARG_MINUTE = "arg_minute"
+
+        fun newInstance(hour: Int, minute: Int): BloggingReminderTimePicker {
+            return BloggingReminderTimePicker().apply {
+                arguments = Bundle().apply {
+                    putInt(ARG_HOUR, hour)
+                    putInt(ARG_MINUTE, minute)
+                }
+            }
+        }
```
It seems that we're not using these arguments anywhere yet. I imagine you're planning to use them to pass the currently selected time to the dialog? If so, I think that's the way to go 👍

ravishanker replied:
Yes, once I merge the FluxC PR

ravishanker replied:
Removed the arguments now, no longer needed as time and hour are available after FluxC update

renanferrari replied:
I believe we should still pass the currently selected time to the dialog. If we don't, the time picker will display a different time from the one that's being displayed on the bottom sheet:


https://user-images.githubusercontent.com/830056/129103685-a9fbb2cd-ab2f-4e20-b06a-1fcd0fea8b4d.mp4

ravishanker replied:
Updated to reflect current time selection on the picker after merging FluxC PR

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersItem.kt:
```diff
@@ -44,5 +52,11 @@ sealed class BloggingRemindersItem(val type: Type) {
 
+    data class TimeItem(
+        val time: UiString,
+        val isInvisible: Boolean = false,
```
It looks like `isInvisible` is not being used anywhere?

ravishanker replied:
Need to seek clarification on this one.  We had discussion if you remember, whether the time selection should always be displayed or only after a day selection is made.  If we decide to default to display always, then this can be removed

ravishanker replied:
Update it now, to be visible only when at least a day is selected

renanferrari replied:
I'm not sure what's the best approach here, but the way we are currently hiding the item looks quite weird IMHO:

https://user-images.githubusercontent.com/830056/129104115-b58e87e0-2520-4219-937a-28fd23ed32b1.mp4

If we want to hide that, we should consider conditionally adding this [`TimeItem`](https://github.com/wordpress-mobile/WordPress-Android/blob/2a90f3d7507fbcd2c95dc4ac9aa7a0009e59b17d/WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/DaySelectionBuilder.kt#L52-L56) to the list instead of hiding its content. Personally, I'd probably just default to always show it, though.

Either way, we should probably ask for a design review here. You can do that by adding the "Needs design review" label and/or pinging a designer yourself.

ravishanker replied:
Yeah, Chris is AFK, will ask next week when he returns.  For now, made TimeItem conditional.  Thanks

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersUiModel.kt:
```diff
@@ -2,4 +2,19 @@ package org.wordpress.android.ui.bloggingreminders
+data class BloggingRemindersUiModel(
+    val siteId: Int,
+    val enabledDays: Set<DayOfWeek> = setOf(),
+    val hour: Int,
+    val minute: Int
```
We should probably use `LocalTime` here instead – it can be constructed pretty easily with `LocalTime.of(hour, minute)` and it would make formatting easier (left a [comment below](https://github.com/wordpress-mobile/WordPress-Android/pull/15128#discussion_r686254015) with an example).

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersUiModel.kt:
```diff
@@ -2,4 +2,19 @@ package org.wordpress.android.ui.bloggingreminders
+    val minute: Int
+) {
+    fun getNotificationTime(): CharSequence {
+        val period = if (hour >= HOUR_12) "PM" else "AM"
+        val hour = if (hour > HOUR_12) hour - HOUR_12 else hour
+        return "$hour:$minute $period"
```
If we use `LocalTime`, we could then have something like:

```kotlin
return LocalTime.of(hour, minute).format(DateTimeFormatter.ofLocalizedTime(FormatStyle.SHORT))
```

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersViewHolder.kt:
```diff
@@ -157,4 +159,14 @@ sealed class BloggingRemindersViewHolder<T : ViewBinding>(protected val binding:
+            ) {
+        fun onBind(item: TimeItem) = with(binding) {
+            uiHelpers.setTextOrHide(timeButton, item.time)
+            timeButton.setOnClickListener { item.onClick.click() }
```
We should probably set the click listener on the whole view, instead of just the `TextView`. Otherwise, the touch target gets really small and easy to miss.

ravishanker replied:
I was wondering the same, should the padding needs to added here around time or just make the whole thing touch target.  I'll change to the whole view for now.

ravishanker replied:
Update touch target to the list item

renanferrari replied:
Making the whole view the target seems the right thing to do 👍 I'd just make sure to add the selectable background to the view, so we can see ripple animation when we touch it.

ravishanker replied:
added selectableItemBackground.  Thanks

renanferrari reviewed (commented):
Thanks for the changes! After testing, it seems that the dialog behavior is still there if we press the back button. I've taken a look at the code and left a few suggestions below. Let me know what you think.

ravishanker committed:
Update for review

refactored time picker handling

ravishanker committed:
Update BloggingRemindersViewModel.kt

Fix ciktlint issue

ravishanker commented on build.gradle:
```diff
@@ -133,7 +133,7 @@ ext {
 
     daggerVersion = '2.29.1'
-    fluxCVersion = '1.23.0'
+    fluxCVersion = '2085-6413bc370255bfebde6e386a05d004cecc81372e'
```
NOTE to self:  Update this to FluxC version after merge

renanferrari reviewed (commented):
Left a few more comments above. Let me know if you need a hand with anything!

renanferrari commented on WordPress/src/main/java/org/wordpress/android/workers/reminder/ReminderScheduler.kt:
```diff
@@ -16,10 +16,12 @@ class ReminderScheduler @Inject constructor(
     val contextProvider: ContextProvider
 ) {
     val workManager by lazy { WorkManager.getInstance(contextProvider.getContext()) }
+    var hour: Int = DEFAUlT_START_HOUR
+    var minute: Int = DEFAULT_START_MINUTE
```
Instead of adding new mutable variables here, we should probably either add a `time` argument to the `schedule` function or refactor `ReminderConfig` to handle `LocalDateTime`s instead of `LocalDate`s. The first option seems simpler.

ravishanker replied:
Updated schedule function with time arguments after merging FluxC PR.  Thanks

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/DaySelectionBuilder.kt:
```diff
@@ -25,14 +27,16 @@ class DaySelectionBuilder
                 EmphasizedText(text),
                 bloggingRemindersModel?.enabledDays?.isEmpty() == true
         )
+        val time = bloggingRemindersModel?.getNotificationTime() ?: "10:00 AM"
```
We shouldn't need to ever set this hardcoded time 🤔

ravishanker replied:
Cleaned up, after merging FluxC PR.   Thanks

renanferrari reviewed (commented):
Sorry, just noticed a few more things!

ravishanker committed:
Refactor and Update after FluxC

Refactored and updated after merging FluxC PR

ravishanker committed:
fix ciktlint issue

ravishanker committed:
Add tracking for time selection

Track time selection feature by adding a selected_time property to the blogging_reminders_scheduled event

ravishanker committed:
fix test

renanferrari reviewed (approved):
Great work!

LGTM :shipit: 

Feel free to merge this after resolving the merge conflicts. I think we can address any design issues separately as well, but it's up to you.

ravishanker committed:
Merge branch 'develop' into feature/15123-blogging-remindertime-selection

ravishanker committed:
Update tracking time to 24 h format

ravishanker committed:
Update BloggingRemindersViewModel.kt

fix lint checks

ravishanker committed:
Merge pull request #15178 from wordpress-mobile/feature/15177-track-events-for-blogging-reminders-v2

Add tracking for time selection

ravishanker committed:
Update build.gradle

with FluxCVersion

osullivanchris commented:
Hey @renanferrari @ravishanker  I had a think about the time display when the user has not selected any days yet. 

What if we display it but disabled until days are selected? Then we follow the same rational that its not interactive before days are selected, but, the UI doesn't jump when it becomes active. WDYT?

The only thing I wasn't sure is if we should display the default time, or no time, before a day is selected.

<img width="898" alt="notification time" src="https://user-images.githubusercontent.com/50576199/129863815-d29bf836-33b2-4d8f-a869-990c87cbf2be.png">

I also thought Ravi had a nice idea about using a button style for the time to make it more clear that its interactive. Here's a quick mock. I think it looks pretty nice but I'm not sure if its ok to put a text button in a list item like this? 

<img width="1065" alt="alt" src="https://user-images.githubusercontent.com/50576199/129864120-bab8d292-81c2-4171-9d1e-94db637c1dd0.png">

osullivanchris commented:
As discussed elsewhere, we have good reason for 'hiding' rather than disabling the time initially. So we can ignore my above suggestion for now!
